{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "challenging-empire",
   "metadata": {},
   "source": [
    "# hiseq.align\n",
    "\n",
    "\n",
    "The standard port for aligner\n",
    "\n",
    "## 1 Requirements  \n",
    "\n",
    "\n",
    "### 1.1 Arguments\n",
    "\n",
    "  - input: str \n",
    "  - index: str\n",
    "  - outdir: pwd/str\n",
    "  - unique: True/False\n",
    "  - aligner: (guess aligner) \n",
    "  - smp_name: None/str\n",
    "  - index_name: None/str\n",
    "  - threads: 4/int\n",
    "  - overwrite: True/False\n",
    "  - extra_para: (in dict, toml)\n",
    "\n",
    "\n",
    "### 1.2 Output\n",
    "\n",
    "  - config.toml\n",
    "  - bam  \n",
    "  - log  \n",
    "  - stat \n",
    "  - cmd.sh \n",
    "  - unmap  \n",
    "  - flagstat \n",
    "\n",
    "\n",
    "### 1.3 Directory structure \n",
    "\n",
    "  - outdir/smp_name/index_name/files\n",
    "\n",
    "\n",
    "\n",
    "## 2. Interpreter \n",
    "\n",
    "\n",
    "\n",
    "### 2.1 Functions \n",
    "\n",
    "  - unique|multiple, determine the aligner \n",
    "  \n",
    "\n",
    "### 2.2 Output\n",
    "\n",
    "  - required arguments for `Align` \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## 3 Aligner\n",
    "\n",
    "\n",
    "### 3.1 Bowtie   \n",
    "\n",
    "```\n",
    "# common: -p 4 --mm\n",
    "# default: k=1\n",
    "bowtie -k 1 -S -v 2 --best --un unmap.fq -x index se.fq 1> align.sam 2> align.log\n",
    "bowtie -k 1 -S -v 2 --best --un unmap.fq -x index -1 pe_1.fq -2 pe_2.fq 1> align.sam 2> align.log\n",
    "\n",
    "# unique\n",
    "bowtie -m 1 -S -v 2 --best --un unmap.fq -x index se.fq 1> align.sam 2> align.log\n",
    "bowtie -m 1 -S -v 2 --best --un unmap.fq -x index -1 pe_1.fq -2 pe_2.fq 1> align.sam 2> align.log\n",
    "```\n",
    "\n",
    "### 3.2 Bowtie2 \n",
    "\n",
    "```\n",
    "# common: -p 4 --mm\n",
    "# default:\n",
    "bowtie2 --very-sensitive --no-unal -x index -U se.fq --un unmap.fq 1> align.sam 2> align.log\n",
    "bowtie2 --very-sensitive --no-unal --no-mixed --no-discordant --un-conc unmap.fq 1> align.sam 2> align.log\n",
    "\n",
    "# unique\n",
    "# https://www.biostars.org/p/101533/#101537, -q 5,10\n",
    "samtools view -bhs -q 10 align.bam \n",
    "\n",
    "\n",
    "# extra parameters\n",
    "-I minimum fragment length (0) \n",
    "-X maximum fragment length (500) \n",
    "```\n",
    "\n",
    "\n",
    "### 3.3 STAR \n",
    "\n",
    "```\n",
    "# common\n",
    "# default:\n",
    "STAR \\\n",
    "    --genomeLoad NoSharedMemory \\\n",
    "    --runMode alignReads \\\n",
    "    --genomeDir index \\\n",
    "    --readFilesIn pe_1 pe_2 \\\n",
    "    --readFilesCommand zcat \\\n",
    "    --outFileNamePrefix smp_name \\\n",
    "    --runThreadN 4 \\\n",
    "    --limitBAMsortRAM 10000000000 \\\n",
    "    --outSAMtype BAM SortedByCoordinate \\\n",
    "    --outFilterMismatchNoverLmx 0.07 \\\n",
    "    --seedSearchStartLmax 20 \\\n",
    "    --outReadsUnmapped Fastx \n",
    "\n",
    "# unique:\n",
    "--outFilterMultimapNmax 1\n",
    "\n",
    "# small genome (default: 50)\n",
    "--seedPerWindowNmax 5\n",
    "\n",
    "\n",
    "# For sharing memory in STAR\n",
    "# by Devon Ryan: https://www.biostars.org/p/260069/#260077 \n",
    "# by Dobin: https://github.com/alexdobin/STAR/pull/26\n",
    "# --genomeLoad\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Updates:\n",
    "\n",
    "1. style the code with PEP8: https://www.python.org/dev/peps/pep-0008/\n",
    "\n",
    "in brief:\n",
    "\n",
    "1 indentation: 4 spaces  \n",
    "    1.1 Aligned with opening delimiter\n",
    "2 Maximum Line Length: 79 characters \n",
    "    2.1 Multi lines, operator at end\n",
    "3 Blank Lines\n",
    "    3.1 Two blank lines surround top-level function/class  \n",
    "    3.2 Single blank line, in class  \n",
    "    3.3 \n",
    "4 White spaces  \n",
    "    4.1 no spaces after/before brackets/braces  \n",
    "    4.2 single space on both sides of binary operator   \n",
    "    4.3 no spaces around '=' for keyword argument  \n",
    "\n",
    "5 Naming styles\n",
    "\n",
    "5.1 lower_case_with_underscores, \n",
    "\n",
    "6 Package and Module Names \n",
    "\n",
    "6.1 Modules should have short, all-lowercase names\n",
    "\n",
    "7 Class names should normally use the CapWords convention. \n",
    "\n",
    "8 Function and Variable Names,  lowercase, with words separated by underscores as necessary to improve readability.\n",
    "\n",
    "9 Use `is`, `is not`\n",
    "\n",
    "10 Use ''.startswith(), ''.endswith()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "everyday-priority",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import aligner_index\n",
    "from aligner_index import *\n",
    "from utils import *\n",
    "\n",
    "args = {\n",
    "    'genome': 'mm10',\n",
    "    'aligner': 'bowtie',\n",
    "    # 'spikein': 'dm3',\n",
    "    # 'to_rRNA': True,\n",
    "    'to_chrM': True,\n",
    "    # 'to_MT_trRNA': True,\n",
    "    'verbose': True,\n",
    "}\n",
    "\n",
    "# check_index_args(**args)\n",
    "\n",
    "# fetch_index('dm6', 'chrM', 'bowtie')\n",
    "\n",
    "# x = '/home/wangming/data/genome/dm6/bowtie_index/chrM'\n",
    "# aligner = 'bowtie'\n",
    "\n",
    "# p = AlignIndex(x, aligner)\n",
    "\n",
    "# p.is_bowtie_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "numeric-gather",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-03-17T04:37:44.703818Z",
     "iopub.status.busy": "2021-03-17T04:37:44.703331Z",
     "iopub.status.idle": "2021-03-17T04:37:46.723062Z",
     "shell.execute_reply": "2021-03-17T04:37:46.722356Z",
     "shell.execute_reply.started": "2021-03-17T04:37:44.703769Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2021-03-17 12:37:46 INFO] Bowtie2() skipped, file exists: /data/yulab/wangming/work/wmlib/hiseq/hiseq/aligner/test/pe_control_rep1/chrM/pe_control_rep1.bam\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('/data/yulab/wangming/work/wmlib/hiseq/hiseq/aligner/test/pe_control_rep1/chrM/pe_control_rep1.bam',\n",
       " '/data/yulab/wangming/work/wmlib/hiseq/hiseq/aligner/test/pe_control_rep1/chrM/pe_control_rep1.unmap.1.fastq',\n",
       " '/data/yulab/wangming/work/wmlib/hiseq/hiseq/aligner/test/pe_control_rep1/chrM/pe_control_rep1.unmap.2.fastq')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from bowtie2 import *\n",
    "\n",
    "args = {\n",
    "    'fq1': '/data/yulab/wangming/work/devel_pipeline/hiseq/rnaseq/data/pe_control_rep1_1.fq.gz',\n",
    "    'fq2': '/data/yulab/wangming/work/devel_pipeline/hiseq/rnaseq/data/pe_control_rep1_2.fq.gz',\n",
    "    'index': '/home/wangming/data/genome/dm6/bowtie2_index/chrM',\n",
    "    'outdir': 'test',\n",
    "    'genome': 'dm6',\n",
    "    'threads': 24,\n",
    "    'unique_only': True,\n",
    "}\n",
    "\n",
    "# a = BowtieConfig(**args)\n",
    "# check_fx_args(args['fq1'], args['fq2'])\n",
    "\n",
    "\n",
    "a = Bowtie2(**args)\n",
    "a.run()\n",
    "\n",
    "\n",
    "\n",
    "# x = 'test/pe_control_rep1/chrM/pe_control_rep1.align.log'\n",
    "# parse_bowtie2(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nutritional-steel",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from bowtie2 import *\n",
    "\n",
    "help(check_file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
