---
title: "ATAC-seq report"
author: "Ming Wang"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: tango
    toc: yes
    toc_float:
      collapsed: no
    keep_md: true
  word_document:
    toc: yes
  pdf_document:
    toc: yes
params:
  input_dir: ""
---

```{r setup, include=FALSE}
suppressPackageStartupMessages(library(hiseqr))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(glue))
suppressPackageStartupMessages(library(ggcor))

knitr::opts_chunk$set(fig.width  = 12, 
                      fig.height = 8, 
                      fig.path   = "Figures/",
                      echo       = FALSE,
                      cache      = FALSE,
                      prompt     = FALSE,
                      tidy       = FALSE,
                      comment    = NA,
                      message    = FALSE,
                      warning    = FALSE,
                      eval       = T,
                      rownames.print = FALSE)
```

```{r samples, eval = T}
align_dir <- file.path(qc_dir, "..", "align")
flist     <- list.files(align_dir, "*align.txt", T, T, T)
df        <- alignStat(flist)
```

## Summary

```{r summary, results="asis"}
mito_pct <- round(df$mito.pct * 100, 2)
summary <- glue::glue("Sample name: {df$sample}, \\n 
                      input total {df$total} reads; \\n
                      uniquely mapped to reference genome {df$dm6.u} reads; \\n
                      contains {mito_pct}% Mitochondrial DNA reads.")
print(summary)
```

## Results

### 1 Table1. Mito Percentage

```{r table1_mito_pct}
knitr::kable(df)
```

### 2 Figure 1. Number of mapped reads

```{r figure1_align_reads, out.width = "100%", fig.height = 4}
df2 <- dplyr::select(df, -mito.pct)
alignPlot(df2)
```

### 3 Figure2. Number of peaks

```{r figure2_peaks, out.width = "100%", fig.height = 4}
## total reads
df3 <- df %>%
  dplyr::select(sample, dm6.u) %>%
  dplyr::rename(count = dm6.u) %>%
  mutate(sample = factor(sample, levels = rev(sample)),
         count  = round(count / 1e6, 1))

## total peaks
peak_dir <- file.path(qc_dir, "..", "peak")
peak_files <- list.files(peak_dir, "*narrowPeak", T, T, T)
gr_list <- lapply(peak_files, narrowPeakReader)

df4 <- data.frame(sample = gsub("_peaks.narrowPeak", "", basename(peak_files)),
                 count  = sapply(gr_list, length)) %>%
  mutate(sample = factor(sample, levels = rev(sample)))

p1 <- barplotCount(df3, TRUE) + 
  ggtitle("Unique reads on genome") +
  ylab("Million of reads")

p2 <- barplotCount(df4, TRUE) + 
  ggtitle("Number of peaks") +
  theme(axis.title.y = element_blank(),
        axis.text.y = element_blank())

p <- cowplot::plot_grid(p1, p2, ncol = 2, rel_widths = c(1, 0.6), 
                        labels = "AUTO")

print(p)
```

### 4. Figure3. Fragment length

The insert size distribution of sequenced fragments from human chromatin had clear periodicity of approximately 200 bp, suggesting many fragments are protected by integer multiples of nucleosomes.

```{r figure3_fragment_length, out.width = "100%", fig.height = 4}
flist <- list.files(qc_dir, "length_distribution.txt", T, T, T)
df <- fragReader(flist)
p  <- fragPlot(df)
print(p)
```

### 5. TSS enrichment

**to-do**



