{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "rocky-phase",
   "metadata": {},
   "source": [
    "## hiseq - trackhub \n",
    "\n",
    "> How to create `trackhub` for `bigWig` and `bigBed` files? \n",
    "\n",
    "In this documetation, will demostrate the details, using the sub-module of `hiseq run_trackhub`\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### 1. Install `hiseq` and `trackhub` packages\n",
    "\n",
    "\n",
    "```shell\n",
    "$ conda install -c bioconda trackhub\n",
    "\n",
    "# get the hiseq source code \n",
    "$ git clone {repo}/hiseq \n",
    "$ cd hiseq\n",
    "$ python setup.py install --user \n",
    "```\n",
    "\n",
    "> maybe, require some dependencies: ...\n",
    "\n",
    "The following message indicate the installation is OK: \n",
    "\n",
    "```shell\n",
    "$ hiseq\n",
    "usage:  hiseq <command> [<args>]\n",
    "\n",
    "    The most commonly used sub-commands are:\n",
    "\n",
    "        atac         ATACseq pipeline\n",
    "        rnaseq       RNAseq pipeline\n",
    "        rnaseq2      RNAseq pipeline, simplify version\n",
    "        chipseq      ChIPseq pipeline\n",
    "        cnr          CUN&RUN pipeline\n",
    "\n",
    "        demx         Demultiplexing reads (P7, barcode)\n",
    "        qc           quality control, fastqc\n",
    "        trim         trim adapters, low-quality bases, ...\n",
    "        align        Align fastq/a files to reference genome\n",
    "        quant        Count genes/features\n",
    "        peak         Call peaks using MACS2\n",
    "        motif        Check motifs from a BED/fasta file\n",
    "        report       Create a report to the above commands\n",
    "        go           Run GO analysis on geneset\n",
    "        deseq_pair   Run RNAseq compare\n",
    "\n",
    "        run_trackhub Generate the trackhub urls\n",
    "        fragsize     Fragment size of PE alignments\n",
    "        bam2cor      Correlation between bam files\n",
    "        bam2bw       Convert bam to bigWig\n",
    "        peak2idr     Calculate IDR for multiple Peaks\n",
    "        bed2overlap  Calculate the overlap between bed intervals\n",
    "        sample       Sample fastq file\n",
    "        download     Download files\n",
    "\n",
    "hiseq: error: the following arguments are required: command\n",
    "\n",
    "```\n",
    "\n",
    "Check the submodule `run_trackhub` in `hiseq`:\n",
    "\n",
    "```shell\n",
    "$ hiseq run_trackhub\n",
    "################################################################################\n",
    "# Mini-tutorial [run_trackhub]                                                 #\n",
    "# 1. Generating template files:                                                #\n",
    "# $ python run_trackhub --demo                                                 #\n",
    "# ('     config.yaml : config_template.yaml',)                                 #\n",
    "# ('  subgroups.yaml : subgroups_template.yaml',)                              #\n",
    "#                                                                              #\n",
    "# 2. Move the YAML files to {data_dir}                                         #\n",
    "# $ mv subgroups.yaml {data_dir}/subgroups.yaml                                #\n",
    "# $ mv config.yaml {data_dir}/config.yaml                                      #\n",
    "#                                                                              #\n",
    "# Update the YAML files, according to your data                                #\n",
    "# Required fields - config.yaml                                                #\n",
    "#   - data_dir        # absolute path                                          #\n",
    "#   - genome          # dm6                                                    #\n",
    "#   - label_rm_list   # string, removed from label                             #\n",
    "#   - hub_name        # RNAseq_piwi                                            #\n",
    "#   - remote_dir      # see http_root_dir                                      #\n",
    "#   - position        # chr2L:1-1000                                           #\n",
    "#   - http_root_alias # /upload, see: /var/apache2/sites-available/            #\n",
    "#   - http_root_dir   # /data/public/upload, as above                          #\n",
    "#   - http_root_url   # null, parse IP of HTTP server                          #\n",
    "#                                                                              #\n",
    "# Required fields - subgroups.yaml                                             #\n",
    "#   - mapping                                                                  #\n",
    "#                                                                              #\n",
    "# 3. Generating trackhub files                                                 #\n",
    "# $ hiseq run_trackhub -c {data_dir}/config.yaml                               #\n",
    "#                                                                              #\n",
    "# 4. Find the hub.txt file                                                     #\n",
    "# {remote_dir}/hub_name/{hub_name}_hub.txt                                     #\n",
    "################################################################################\n",
    "\n",
    "```\n",
    "\n",
    "\n",
    "## 2. Getting started\n",
    "\n",
    "\n",
    "### 2.1. Prepare `track files` \n",
    "\n",
    "The file should be organized in the following structure:\n",
    "\n",
    "+ Single track\n",
    "\n",
    "```shell\n",
    "data\n",
    "  |--bw_files\n",
    "  |--bb_files\n",
    "\n",
    "```\n",
    "\n",
    "+ Multiple tracks\n",
    "\n",
    "```shell\n",
    "data\n",
    "  |--sub_data01\n",
    "  |     |--bw_files\n",
    "  |     |--bb_files\n",
    "  |--sub_data02\n",
    "  |     |--bw_files\n",
    "  |     |--bb_files\n",
    "  ...\n",
    "``` \n",
    "\n",
    "Here are supported file formats (more formats in plan)\n",
    "\n",
    "`bigWig` - filename should be named: `*.bigWig` or `*.bw` (required)        \n",
    "`bigBed` - filename should be named: `*.bigBed` or `*.bb`, (optional)\n",
    "\n",
    "see documentation on UCSC for details about the formats: [bigWig](https://genome.ucsc.edu/goldenpath/help/bigWig.html) and [bigBed](https://genome.ucsc.edu/goldenPath/help/bigBed.html)    \n",
    "\n",
    "\n",
    "### 2.2 Generate config files\n",
    "\n",
    "\n",
    "Run the command: \n",
    "\n",
    "```\n",
    "$ hiseq run_trackhub\n",
    "$ ls *.yaml\n",
    "config_template.yaml  subgroups_template.yaml\n",
    "```\n",
    "\n",
    "two template files generated in current working dir: `subgroups_template.yaml`, `config_template.yaml`. \n",
    "\n",
    "The config files are required for down-stream program.\n",
    "\n",
    "`subgroups.yaml` - define the dimX, dimY, ..., to define the subgroups for each sample\n",
    "`config.yaml`    - main arguments for the program    \n",
    "\n",
    "**Attention**\n",
    "\n",
    "  - 1. `subgroups.yaml` should be in the same directory of `bw_files/`.     \n",
    "  - 2. `config.yaml` saved in `data/` directory    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "broken-hampton",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
