{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "detailed-military",
   "metadata": {},
   "outputs": [],
   "source": [
    "## test CnrR1\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import sys\n",
    "from cnr_r1 import CnrR1\n",
    "from hiseq.utils.file import list_file\n",
    "\n",
    "\n",
    "data_dir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/data/fruitfly/'\n",
    "\n",
    "fq1 = list_file(data_dir, '*Piwi*K4*1.fq.gz')\n",
    "fq2 = list_file(data_dir, '*Piwi*K4*2.fq.gz')\n",
    "outdir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/output/fruitfly'\n",
    "\n",
    "args = {\n",
    "    'fq1': fq1[1],\n",
    "    'fq2': fq2[1],\n",
    "    'genome': 'dm6',\n",
    "    'outdir': outdir,\n",
    "    'threads': 12,\n",
    "    'parallel_jobs': 2\n",
    "}\n",
    "\n",
    "CnrR1(**args).run()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "elect-shame",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-06-15T15:17:40.776851Z",
     "iopub.status.busy": "2021-06-15T15:17:40.776258Z",
     "iopub.status.idle": "2021-06-15T15:17:42.406564Z",
     "shell.execute_reply": "2021-06-15T15:17:42.405822Z",
     "shell.execute_reply.started": "2021-06-15T15:17:40.776713Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/raw_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1_1.fq.gz\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/raw_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1_2.fq.gz\n",
      "[2021-06-15 23:17:41 INFO] trim() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/clean_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1_1.fq.gz\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/clean_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1_1.fq.gz\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/clean_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1_2.fq.gz\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/clean_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.trim.json\n",
      "[2021-06-15 23:17:41 INFO] atac_align_genome() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/bam_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.bam\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/bam_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.raw.bam\n",
      "[2021-06-15 23:17:41 WARNING] symlink_file() failed, src not vaild: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/align/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.align.stat\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/align/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.align.json\n",
      "[2021-06-15 23:17:41 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/align/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.align.flagstat\n",
      "[2021-06-15 23:17:41 INFO] rmdup() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/bam_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.bam\n",
      "--------------------------------------------------------------------------------\n",
      "          Date : 2021-06-15 23:17:41\n",
      "       program : hiseq.cnr.CallPeak\n",
      "        config : /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak/config.yaml\n",
      "    peakcaller : macs2\n",
      "        genome : dm6\n",
      "        outdir : /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak\n",
      "            ip : /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/bam_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.bam\n",
      "         input : None\n",
      "   genome_size : dm\n",
      "genome_size_file : /home/wangming/data/genome/dm6/bigZips/dm6.fa.chrom.sizes\n",
      "      keep_tmp : False\n",
      "     overwrite : False\n",
      "    hiseq_type : callpeak_r1\n",
      "--------------------------------------------------------------------------------\n",
      "[2021-06-15 23:17:41 INFO] run_macs2() skipped, file exists:/data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1_peaks.narrowPeak\n",
      "--------------------------------------------------------------------------------\n",
      "          Date : 2021-06-15 23:17:42\n",
      "       program : hiseq.cnr.CallPeak\n",
      "        config : /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak/config.yaml\n",
      "    peakcaller : seacr\n",
      "        genome : dm6\n",
      "        outdir : /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak\n",
      "            ip : /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/bam_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.bam\n",
      "         input : None\n",
      "   genome_size : dm\n",
      "genome_size_file : /home/wangming/data/genome/dm6/bigZips/dm6.fa.chrom.sizes\n",
      "      keep_tmp : False\n",
      "     overwrite : False\n",
      "    hiseq_type : callpeak_r1\n",
      "--------------------------------------------------------------------------------\n",
      "[2021-06-15 23:17:42 INFO] run_seacr() skipped, file exists:/data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.top0.01.stringent.bed\n",
      "[2021-06-15 23:17:42 INFO] Bam2bw() skipped, file exists, ..., /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/bw_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.bigWig\n",
      "[2021-06-15 23:17:42 INFO] qc_lendist() skipped: file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/qc/02.length_distribution.fragsize.csv\n",
      "[2021-06-15 23:17:42 INFO] qc_frip() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/qc/03.FRiP.json\n",
      "[2021-06-15 23:17:42 INFO] qc_tss() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/qc/04.tss_enrich.png\n",
      "[2021-06-15 23:17:42 INFO] qc_genebody_enrich() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/qc/05.genebody_enrich.png\n",
      "[2021-06-15 23:17:42 INFO] report() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/report/HiSeq_report.html\n",
      "[2021-06-15 23:17:42 WARNING] merge() skipped, Only 1 replicate detected\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/bam_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG.bam\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/bw_files/CnR_Nos_DaGal4XshPiwi_6-8h_IgG.bigWig\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/peak/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_peaks.narrowPeak\n",
      "[2021-06-15 23:17:42 WARNING] symlink_file() failed, src not vaild: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/peak/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1.stringent.bed\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/peak/CnR_Nos_DaGal4XshPiwi_6-8h_IgG.top0.01.stringent.bed\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/bam_files/scale.json\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/clean_data/CnR_Nos_DaGal4XshPiwi_6-8h_IgG.trim.json\n",
      "[2021-06-15 23:17:42 INFO] symlink_file() skipped, dest exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/align/CnR_Nos_DaGal4XshPiwi_6-8h_IgG.bowtie2.json\n",
      "[2021-06-15 23:17:42 WARNING] symlink_file() failed, src not vaild: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1/qc/frip_files\n",
      "[2021-06-15 23:17:42 INFO] report() skipped, file exists: /data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/report/HiSeq_report.html\n"
     ]
    }
   ],
   "source": [
    "## test CnrRn\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import sys\n",
    "from cnr_r1 import CnrR1\n",
    "from cnr_rn import CnrRn\n",
    "from hiseq.utils.file import list_file\n",
    "from hiseq.utils.utils import read_hiseq, Config, list_hiseq_file\n",
    "from utils import qc_frip\n",
    "\n",
    "# data_dir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/data/fruitfly/'\n",
    "\n",
    "\n",
    "# fq1 = list_file(data_dir, '*Piwi*IgG*1.fq.gz')\n",
    "# fq2 = list_file(data_dir, '*Piwi*IgG*2.fq.gz')\n",
    "\n",
    "# outdir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/output/fruitfly'\n",
    "\n",
    "# args = {\n",
    "#     'fq1': fq1,\n",
    "#     'fq2': fq2,\n",
    "#     'genome': 'dm6',\n",
    "#     'outdir': outdir,\n",
    "#     'threads': 12,\n",
    "#     'parallel_jobs': 4\n",
    "# }\n",
    "\n",
    "# CnrRn(**args).run()\n",
    "\n",
    "\n",
    "c = '/data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG/config/config.yaml'\n",
    "args = Config().load(c)\n",
    "a = CnrRn(**args)\n",
    "# a.run_pipe_r1()\n",
    "a.run()\n",
    "\n",
    "# rep_dir = '/data/yulab/wangming/work/yu_2021/projects/20210421_lxh_CnR/results/CnR_v2/CnR_Nos_DaGal4XshPiwi_6-8h_IgG_rep1'\n",
    "# k = 'bam'\n",
    "# list_hiseq_file(rep_dir, k, 'r1')\n",
    "\n",
    "\n",
    "\n",
    "# x = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/output/fruitfly/CUT_and_RUN_Piwi_RNAi_embryo_3_4h_H3K4me3/'\n",
    "# pd = read_hiseq(x)\n",
    "# qc_frip(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "common-purpose",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test CnrRx # design\n",
    "\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import sys\n",
    "from cnr_r1 import CnrR1\n",
    "from cnr_rn import CnrRn\n",
    "from cnr_rx import CnrRx\n",
    "from hiseq.utils.file import list_file\n",
    "from hiseq.utils.utils import read_hiseq, print_dict\n",
    "from utils import qc_frip\n",
    "data_dir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/data/fruitfly/'\n",
    "\n",
    "ip_fq1 = list_file(data_dir, '*White*K4*1.fq.gz')\n",
    "ip_fq2 = list_file(data_dir, '*White*K4*2.fq.gz')\n",
    "input_fq1 = list_file(data_dir, '*White*IgG*1.fq.gz')\n",
    "input_fq2 = list_file(data_dir, '*White*IgG*2.fq.gz')\n",
    "\n",
    "outdir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/output/fruitfly'\n",
    "\n",
    "args = {\n",
    "    'ip_fq1': ip_fq1,\n",
    "    'ip_fq2': ip_fq2,\n",
    "    'input_fq1': input_fq1,\n",
    "    'input_fq2': input_fq2,\n",
    "    'genome': 'dm6',\n",
    "    'outdir': outdir,\n",
    "    'threads': 12,\n",
    "    'parallel_jobs': 4,\n",
    "    'gene_bed': '/home/wangming/data/genome/dm6/annotation_and_repeats/dm6.ensembl.bed',\n",
    "}\n",
    "\n",
    "CnrRx(**args).run()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "searching-fundamental",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T04:37:02.953908Z",
     "iopub.status.busy": "2021-05-31T04:37:02.953365Z",
     "iopub.status.idle": "2021-05-31T04:37:02.991456Z",
     "shell.execute_reply": "2021-05-31T04:37:02.990721Z",
     "shell.execute_reply.started": "2021-05-31T04:37:02.953858Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pct</th>\n",
       "      <th>reads_in_clade</th>\n",
       "      <th>reads_in_tax</th>\n",
       "      <th>code</th>\n",
       "      <th>taxid</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1988</th>\n",
       "      <td>0.02</td>\n",
       "      <td>825</td>\n",
       "      <td>617</td>\n",
       "      <td>G</td>\n",
       "      <td>2132</td>\n",
       "      <td>Spiroplasma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2418</th>\n",
       "      <td>1.02</td>\n",
       "      <td>40737</td>\n",
       "      <td>354</td>\n",
       "      <td>G</td>\n",
       "      <td>570</td>\n",
       "      <td>Klebsiella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2928</th>\n",
       "      <td>0.05</td>\n",
       "      <td>1834</td>\n",
       "      <td>311</td>\n",
       "      <td>G</td>\n",
       "      <td>286</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>0.08</td>\n",
       "      <td>3298</td>\n",
       "      <td>265</td>\n",
       "      <td>G</td>\n",
       "      <td>1883</td>\n",
       "      <td>Streptomyces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1310</th>\n",
       "      <td>0.02</td>\n",
       "      <td>708</td>\n",
       "      <td>236</td>\n",
       "      <td>G</td>\n",
       "      <td>1763</td>\n",
       "      <td>Mycobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5925</th>\n",
       "      <td>0.01</td>\n",
       "      <td>235</td>\n",
       "      <td>235</td>\n",
       "      <td>G</td>\n",
       "      <td>160798</td>\n",
       "      <td>Marinitoga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6563</th>\n",
       "      <td>0.00</td>\n",
       "      <td>192</td>\n",
       "      <td>188</td>\n",
       "      <td>G</td>\n",
       "      <td>181083</td>\n",
       "      <td>Chlorovirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>187</th>\n",
       "      <td>0.35</td>\n",
       "      <td>13945</td>\n",
       "      <td>138</td>\n",
       "      <td>G</td>\n",
       "      <td>1386</td>\n",
       "      <td>Bacillus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4335</th>\n",
       "      <td>0.00</td>\n",
       "      <td>132</td>\n",
       "      <td>130</td>\n",
       "      <td>G</td>\n",
       "      <td>34008</td>\n",
       "      <td>Rhodovulum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4181</th>\n",
       "      <td>0.01</td>\n",
       "      <td>308</td>\n",
       "      <td>123</td>\n",
       "      <td>G</td>\n",
       "      <td>773</td>\n",
       "      <td>Bartonella</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       pct  reads_in_clade  reads_in_tax code   taxid  \\\n",
       "1988  0.02             825           617    G    2132   \n",
       "2418  1.02           40737           354    G     570   \n",
       "2928  0.05            1834           311    G     286   \n",
       "948   0.08            3298           265    G    1883   \n",
       "1310  0.02             708           236    G    1763   \n",
       "5925  0.01             235           235    G  160798   \n",
       "6563  0.00             192           188    G  181083   \n",
       "187   0.35           13945           138    G    1386   \n",
       "4335  0.00             132           130    G   34008   \n",
       "4181  0.01             308           123    G     773   \n",
       "\n",
       "                               name  \n",
       "1988                    Spiroplasma  \n",
       "2418                     Klebsiella  \n",
       "2928                    Pseudomonas  \n",
       "948                    Streptomyces  \n",
       "1310                  Mycobacterium  \n",
       "5925                     Marinitoga  \n",
       "6563                    Chlorovirus  \n",
       "187                        Bacillus  \n",
       "4335                     Rhodovulum  \n",
       "4181                     Bartonella  "
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "def parse_report(in_report, out_stat, topN=10, tax_level='G'):\n",
    "        \"\"\"\n",
    "        Parameters\n",
    "        ----------\n",
    "        in_report: str\n",
    "            The --report output of kraken2\n",
    "            \n",
    "        out_stat: str\n",
    "            The file saving the simplified report, by specific tax level\n",
    "            \n",
    "        topN: int\n",
    "            The top N taxon, default: 100\n",
    "            \n",
    "        tax_level: str\n",
    "            The specific tax level to show, default: ['G'],\n",
    "            options: [D, K, P, C, O, F, G, S]\n",
    "        \n",
    "        parse Kraken2 report (--report)\n",
    "        Inspect the output of kraken2 report file\n",
    "        pandas table\n",
    "        \n",
    "        col-1: Percentage of fragments covered by the clade rooted at this taxon\n",
    "        col-2: Number of fragments covered by the clade rooted at this taxon\n",
    "        col-3: Number of fragments assigned directly to this taxon\n",
    "        col-4: A rank code, indicating (U)nclassified, (R)oot, (D)omain, \n",
    "               (K)ingdom, (P)hylum, (C)lass, (O)rder, (F)amily, (G)enus,\n",
    "               or (S)pecies. Taxa that are not at any of these 10 ranks have \n",
    "               a rank code that is formed by using the rank code of the closest\n",
    "               ancestor rank with a number indicating the distance from that rank.\n",
    "               E.g., \"G2\" is a rank code indicating a taxon is between genus and \n",
    "               species and the grandparent taxon is at the genus rank.\n",
    "        col-5: NCBI taxonomic ID number\n",
    "        col-6: Indented scientific name\n",
    "        \n",
    "        # choose the top species, by col-3, read directly in taxon\n",
    "        \"\"\"\n",
    "        s = ['pct', 'reads_in_clade', 'reads_in_tax', 'code', 'taxid', 'name']\n",
    "        try:\n",
    "            df1 = pd.read_csv(in_report, '\\t', names=s)\n",
    "        except IOError as e:\n",
    "            log.error(e)\n",
    "            return None\n",
    "        # choose Geneus (G); root, unclassified\n",
    "        df_un = df1.loc[df1['code'] == 'U'] # unclassified\n",
    "        df_root = df1.loc[df1['code'] == 'R'] # root\n",
    "        df_tax = df1.loc[df1['code'].isin([tax_level]), ] # eg: G\n",
    "        # choose top ranked taxon\n",
    "        df_tax = df_tax.sort_values(['reads_in_tax'], ascending=False)\n",
    "        # sub-sample\n",
    "        df2 = df_tax.loc[0:topN,]\n",
    "#         # combine\n",
    "#         df = pd.concate([df_un, df2])\n",
    "        ## remove white spaces\n",
    "#         df['name'] = df['name'].str.strip()\n",
    "#         # add sample name\n",
    "#         df['sample'] = file_prefix(x) # prefix\n",
    "#         # get total\n",
    "#         n_hit = df_root.loc[:, 'reads_in_clade']\n",
    "#         n_unhit = df_un.loc[:, 'reads_in_clade']\n",
    "#         # adt pct\n",
    "#         df['reads_hit'] = int(n_hit) # root\n",
    "#         df['reads_total'] = int(n_hit) + int(n_unhit) # root + unclassified\n",
    "#         df['hit_pct'] = df['reads_in_tax'] / int(n_hit) * 100\n",
    "#          # sub-sample\n",
    "#         df3 = df.loc[:,['sample', 'name', 'reads_in_tax', 'hit_pct', 'reads_hit', 'reads_total']]        \n",
    "#         # custome options\n",
    "#         with pd.option_context('display.expand_frame_repr', False, 'display.max_colwidth', 20):\n",
    "#             print(df3)\n",
    "#         # save to file\n",
    "#         df3.to_csv(out_stat, sep='\\t', index=False)\n",
    "#         return df3\n",
    "\n",
    "in_report = '/data/yulab/wangming/work/devel_pipeline/hiseq/qc/aaaa.kraken2.report'\n",
    "out_stat = '/data/yulab/wangming/work/devel_pipeline/hiseq/qc/aaaa.kraken2.stat'\n",
    "topN = 10\n",
    "# parse_report(in_report, out_stat)\n",
    "tax_level = 'G'\n",
    "s = ['pct', 'reads_in_clade', 'reads_in_tax', 'code', 'taxid', 'name']\n",
    "try:\n",
    "    df1 = pd.read_csv(in_report, '\\t', names=s)\n",
    "except IOError as e:\n",
    "    log.error(e)\n",
    "# choose Geneus (G); root, unclassified\n",
    "df_un = df1.loc[df1['code'] == 'U'] # unclassified\n",
    "df_root = df1.loc[df1['code'] == 'R'] # root\n",
    "df_tax = df1.loc[df1['code'].isin([tax_level]), ] # eg: G\n",
    "# choose top ranked taxon\n",
    "df_tax = df_tax.sort_values(['reads_in_tax'], ascending=False)\n",
    "# sub-sample\n",
    "df2 = df_tax.iloc[0:topN,]\n",
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "attached-adaptation",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T09:05:30.729326Z",
     "iopub.status.busy": "2021-05-31T09:05:30.728868Z",
     "iopub.status.idle": "2021-05-31T09:05:30.735609Z",
     "shell.execute_reply": "2021-05-31T09:05:30.734475Z",
     "shell.execute_reply.started": "2021-05-31T09:05:30.729280Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " all([False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "filled-connectivity",
   "metadata": {},
   "outputs": [],
   "source": [
    "# test CnrRd # design\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "from cnr_rd import CnrRd\n",
    "\n",
    "data_dir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/data/fruitfly/'\n",
    "\n",
    "ip_fq1 = list_file(data_dir, '*Piwi*K4*1.fq.gz')\n",
    "ip_fq2 = list_file(data_dir, '*Piwi*K4*2.fq.gz')\n",
    "\n",
    "input_fq1 = list_file(data_dir, '*Piwi*IgG*1.fq.gz')\n",
    "input_fq2 = list_file(data_dir, '*Piwi*IgG*2.fq.gz')\n",
    "outdir = '/data/yulab/wangming/work/devel_pipeline/hiseq/cnr/output/fruitfly'\n",
    "\n",
    "args = {\n",
    "    'design': 'aaa.toml',\n",
    "    'fq_dir': data_dir,\n",
    "    'ip_fq1': ip_fq1,\n",
    "    'ip_fq2': ip_fq2,\n",
    "    'input_fq1': input_fq1,\n",
    "    'input_fq2': input_fq2\n",
    "}\n",
    "\n",
    "\n",
    "args = {\n",
    "    'design': 'aaa.toml',\n",
    "    'fq_dir': data_dir,\n",
    "    'ip': ['Piwi_RNAi_embryo_3_4h_H3K4me3', 'White_RNAi_embryo_3_4h_H3K4me3'],\n",
    "    'input': ['Piwi_RNAi_embryo_3_4h_IgG'],\n",
    "}\n",
    "\n",
    "a = CnrRd(**args)\n",
    "\n",
    "#a.check_fx_args()\n",
    "df = a.run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unauthorized-nightmare",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from difflib import SequenceMatcher\n",
    "\n",
    "string1 = \"apple pie available\"\n",
    "string2 = \"come have some apple pies\"\n",
    "\n",
    "# match = SequenceMatcher(None, string1, string2).find_longest_match(0, len(string1), 0, len(string2))\n",
    "\n",
    "# print(match)  # -> Match(a=0, b=15, size=9)\n",
    "# print(string1[match.a: match.a + match.size])  # -> apple pie\n",
    "# print(string2[match.b: match.b + match.size])  # -> apple pie\n",
    "\n",
    "# string1[match.a:match.a+match.size]\n",
    "\n",
    "# dir(match)\n",
    "\n",
    "\n",
    "## Temp\n",
    "def find_longest_common_str(s1, s2):\n",
    "    if isinstance(s1, str) and isinstance(s2, str):\n",
    "        m = SequenceMatcher(None, s1, s2) # match\n",
    "        l = m.find_longest_match(0, len(s1), 0, len(s2))\n",
    "        out = s1[l.a: l.a + l.size]\n",
    "    else:\n",
    "        log.error('only support str, got s1={} s2={}'.type(\n",
    "            type(s1).__name__, type(s2).__name__))\n",
    "        out = None\n",
    "    return out\n",
    "\n",
    "\n",
    "\n",
    "find_longest_common_str('abc', 'efgab')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
