{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fourth-calgary",
   "metadata": {},
   "source": [
    "# hiseq.align\n",
    "\n",
    "\n",
    "The standard port for aligner\n",
    "\n",
    "## 1 Requirements  \n",
    "\n",
    "\n",
    "### 1.1 Arguments\n",
    "\n",
    "  - input: str \n",
    "  - index: str\n",
    "  - outdir: pwd/str\n",
    "  - unique: True/False\n",
    "  - aligner: (guess aligner) \n",
    "  - smp_name: None/str\n",
    "  - index_name: None/str\n",
    "  - threads: 4/int\n",
    "  - overwrite: True/False\n",
    "  - extra_para: (in dict, toml)\n",
    "\n",
    "\n",
    "### 1.2 Output\n",
    "\n",
    "  - config.toml\n",
    "  - bam  \n",
    "  - log  \n",
    "  - stat \n",
    "  - cmd.sh \n",
    "  - unmap  \n",
    "  - flagstat \n",
    "\n",
    "\n",
    "### 1.3 Directory structure \n",
    "\n",
    "  - outdir/smp_name/index_name/files\n",
    "\n",
    "\n",
    "\n",
    "## 2. Interpreter \n",
    "\n",
    "\n",
    "\n",
    "### 2.1 Functions \n",
    "\n",
    "  - unique|multiple, determine the aligner \n",
    "  \n",
    "\n",
    "### 2.2 Output\n",
    "\n",
    "  - required arguments for `Align` \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## 3 Aligner\n",
    "\n",
    "\n",
    "### 3.1 Bowtie   \n",
    "\n",
    "```\n",
    "# common: -p 4 --mm\n",
    "# default: k=1\n",
    "bowtie -k 1 -S -v 2 --best --un unmap.fq -x index se.fq 1> align.sam 2> align.log\n",
    "bowtie -k 1 -S -v 2 --best --un unmap.fq -x index -1 pe_1.fq -2 pe_2.fq 1> align.sam 2> align.log\n",
    "\n",
    "# unique\n",
    "bowtie -m 1 -S -v 2 --best --un unmap.fq -x index se.fq 1> align.sam 2> align.log\n",
    "bowtie -m 1 -S -v 2 --best --un unmap.fq -x index -1 pe_1.fq -2 pe_2.fq 1> align.sam 2> align.log\n",
    "```\n",
    "\n",
    "### 3.2 Bowtie2 \n",
    "\n",
    "```\n",
    "# common: -p 4 --mm\n",
    "# default:\n",
    "bowtie2 --very-sensitive --no-unal -x index -U se.fq --un unmap.fq 1> align.sam 2> align.log\n",
    "bowtie2 --very-sensitive --no-unal --no-mixed --no-discordant --un-conc unmap.fq 1> align.sam 2> align.log\n",
    "\n",
    "# unique\n",
    "# https://www.biostars.org/p/101533/#101537, -q 5,10\n",
    "samtools view -bhs -q 10 align.bam \n",
    "\n",
    "\n",
    "# extra parameters\n",
    "-I minimum fragment length (0) \n",
    "-X maximum fragment length (500) \n",
    "```\n",
    "\n",
    "\n",
    "### 3.3 STAR \n",
    "\n",
    "```\n",
    "# common\n",
    "# default:\n",
    "STAR \\\n",
    "    --genomeLoad NoSharedMemory \\\n",
    "    --runMode alignReads \\\n",
    "    --genomeDir index \\\n",
    "    --readFilesIn pe_1 pe_2 \\\n",
    "    --readFilesCommand zcat \\\n",
    "    --outFileNamePrefix smp_name \\\n",
    "    --runThreadN 4 \\\n",
    "    --limitBAMsortRAM 10000000000 \\\n",
    "    --outSAMtype BAM SortedByCoordinate \\\n",
    "    --outFilterMismatchNoverLmx 0.07 \\\n",
    "    --seedSearchStartLmax 20 \\\n",
    "    --outReadsUnmapped Fastx \n",
    "\n",
    "# unique:\n",
    "--outFilterMultimapNmax 1\n",
    "\n",
    "# small genome (default: 50)\n",
    "--seedPerWindowNmax 5\n",
    "\n",
    "\n",
    "# For sharing memory in STAR\n",
    "# by Devon Ryan: https://www.biostars.org/p/260069/#260077 \n",
    "# by Dobin: https://github.com/alexdobin/STAR/pull/26\n",
    "# --genomeLoad\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "Updates:\n",
    "\n",
    "1. style the code with PEP8: https://www.python.org/dev/peps/pep-0008/\n",
    "\n",
    "in brief:\n",
    "\n",
    "1 indentation: 4 spaces  \n",
    "    1.1 Aligned with opening delimiter\n",
    "2 Maximum Line Length: 79 characters \n",
    "    2.1 Multi lines, operator at end\n",
    "3 Blank Lines\n",
    "    3.1 Two blank lines surround top-level function/class  \n",
    "    3.2 Single blank line, in class  \n",
    "    3.3 \n",
    "4 White spaces  \n",
    "    4.1 no spaces after/before brackets/braces  \n",
    "    4.2 single space on both sides of binary operator   \n",
    "    4.3 no spaces around '=' for keyword argument  \n",
    "\n",
    "5 Naming styles\n",
    "\n",
    "5.1 lower_case_with_underscores, \n",
    "\n",
    "6 Package and Module Names \n",
    "\n",
    "6.1 Modules should have short, all-lowercase names\n",
    "\n",
    "7 Class names should normally use the CapWords convention. \n",
    "\n",
    "8 Function and Variable Names,  lowercase, with words separated by underscores as necessary to improve readability.\n",
    "\n",
    "9 Use `is`, `is not`\n",
    "\n",
    "10 Use ''.startswith(), ''.endswith()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "substantial-interference",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import aligner_index\n",
    "from aligner_index import *\n",
    "from utils import *\n",
    "\n",
    "args = {\n",
    "    'genome': 'mm10',\n",
    "    'aligner': 'bowtie',\n",
    "    # 'spikein': 'dm3',\n",
    "    # 'to_rRNA': True,\n",
    "    'to_chrM': True,\n",
    "    # 'to_MT_trRNA': True,\n",
    "    'verbose': True,\n",
    "}\n",
    "\n",
    "# check_index_args(**args)\n",
    "\n",
    "# fetch_index('dm6', 'chrM', 'bowtie')\n",
    "\n",
    "# x = '/home/wangming/data/genome/dm6/bowtie_index/chrM'\n",
    "# aligner = 'bowtie'\n",
    "\n",
    "# p = AlignIndex(x, aligner)\n",
    "\n",
    "# p.is_bowtie_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interracial-cornell",
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from bowtie2 import *\n",
    "from star import *\n",
    "\n",
    "\n",
    "args = {\n",
    "    'fq1': '/data/yulab/wangming/work/devel_pipeline/hiseq/rnaseq/data/pe_control_rep1_1.fq.gz',\n",
    "    # 'fq2': '/data/yulab/wangming/work/devel_pipeline/hiseq/rnaseq/data/pe_control_rep1_2.fq.gz',\n",
    "    'index': '/home/wangming/data/genome/dm6/STAR_index/dm6',\n",
    "    'outdir': 'test',\n",
    "    'genome': 'dm6',\n",
    "    'threads': 24,\n",
    "    'unique_only': True,\n",
    "}\n",
    "\n",
    "# a = BowtieConfig(**args)\n",
    "# check_fx_args(args['fq1'], args['fq2'])\n",
    "\n",
    "\n",
    "# a = Bowtie2(**args)\n",
    "# a.run()\n",
    "\n",
    "\n",
    "a = Star(**args)\n",
    "a.run()\n",
    "\n",
    "\n",
    "# x = 'test/pe_control_rep1/chrM/pe_control_rep1.align.log'\n",
    "# parse_bowtie2(x)\n",
    "\n",
    "\n",
    "# x = '/home/wangming/data/genome/dm6/STAR_index/chrM'\n",
    "# a = AlignIndex(x, 'STAR')\n",
    "# a.is_valid()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "artificial-mistake",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-03-17T19:13:59.756908Z",
     "iopub.status.busy": "2021-03-17T19:13:59.756385Z",
     "iopub.status.idle": "2021-03-17T19:13:59.852744Z",
     "shell.execute_reply": "2021-03-17T19:13:59.852104Z",
     "shell.execute_reply.started": "2021-03-17T19:13:59.756857Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['test/data/pe_control_rep1_1.fq.gz',\n",
       " 'test/data/pe_control_rep2_1.fq.gz',\n",
       " 'test/data/pe_treatment_rep1_1.fq.gz',\n",
       " 'test/data/pe_treatment_rep2_1.fq.gz']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "import fnmatch\n",
    "import logging as log\n",
    "\n",
    "## 1. files and path\n",
    "def listdir(x, full_name=True, recursive=False, include_dir=False):\n",
    "    \"\"\"List all the files within the path\n",
    "    see: list.dir() in R\n",
    "    see answers on :https://stackoverflow.com/a/3207973\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    x : str\n",
    "        List files (dirs) in x\n",
    "        \n",
    "    full_name : bool\n",
    "        Return the fullname of the files/dirs \n",
    "        \n",
    "    recursive : bool\n",
    "        List files/dirs recursively \n",
    "        \n",
    "    include_dir : bool\n",
    "        Return the dirs\n",
    "    \"\"\"\n",
    "    out = []\n",
    "    if isinstance(x, str):\n",
    "        if os.path.isdir(x):\n",
    "            n = 0 # levels\n",
    "            for (root, d, f) in os.walk(x):\n",
    "                dirs = [os.path.join(root, i) for i in d] if full_name else d\n",
    "                files = [os.path.join(root, i) for i in f] if full_name else f\n",
    "                out += files\n",
    "                if include_dir:\n",
    "                    out += dirs\n",
    "                if not recursive:\n",
    "                    break # first level\n",
    "        else:\n",
    "            log.error('listdir() skipped, x not a directory: {}'.format(x))\n",
    "    else:\n",
    "        log.error('listdir() skipped, x expect str, got {}'.format(\n",
    "            type(x).__name__))\n",
    "    return out\n",
    "\n",
    "\n",
    "def listfile(path='.', pattern='*', full_name=True, recursive=False,\n",
    "    include_dir=False):\n",
    "    \"\"\"Search files by the pattern, within directory\n",
    "    fnmatch.fnmatch()\n",
    "\n",
    "    pattern:\n",
    "    *       matches everything\n",
    "    ?       matches any single character\n",
    "    [seq]   matches any character in seq\n",
    "    [!seq]  matches any char not in seq\n",
    "\n",
    "    An initial period in FILENAME is not special.\n",
    "    Both FILENAME and PATTERN are first case-normalized\n",
    "    if the operating system requires it.\n",
    "    If you don't want this, use fnmatchcase(FILENAME, PATTERN).\n",
    "\n",
    "    example:\n",
    "    listfile('./', '*.fq')\n",
    "    \"\"\"\n",
    "    files = listdir(path, full_name, recursive, include_dir)\n",
    "    files = [f for f in files if fnmatch.fnmatch(os.path.basename(f), pattern)]\n",
    "    return sorted(files)\n",
    "\n",
    "\n",
    "x = 'test/data/'\n",
    "listfile(x, '*_1.fq.gz', recursive=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "closed-toronto",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-03-18T11:42:13.697377Z",
     "iopub.status.busy": "2021-03-18T11:42:13.696911Z",
     "iopub.status.idle": "2021-03-18T11:42:13.848537Z",
     "shell.execute_reply": "2021-03-18T11:42:13.847624Z",
     "shell.execute_reply.started": "2021-03-18T11:42:13.697328Z"
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'file_is_gzipped' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-94-54f8e15ff4e6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     69\u001b[0m \u001b[0my\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'test/aaa.fq'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 71\u001b[0;31m \u001b[0mgzip_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m<ipython-input-94-54f8e15ff4e6>\u001b[0m in \u001b[0;36mgzip_file\u001b[0;34m(src, dest, decompress, **kwargs)\u001b[0m\n\u001b[1;32m     42\u001b[0m     \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mflag\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 44\u001b[0;31m         \u001b[0mis_gzipped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfile_is_gzipped\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     45\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdecompress\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     46\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_gzipped\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'file_is_gzipped' is not defined"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import re\n",
    "import shutil\n",
    "import logging as log\n",
    "from utils import *\n",
    "\n",
    "\n",
    "def gzip_file(src, dest=None, decompress=True, **kwargs):\n",
    "    \"\"\"Gzip Compress or Decompress files using gzip module in python\n",
    "    rm, True/False, whether remove old file\n",
    "\n",
    "    # check the src file by extension: .gz\n",
    "    \"\"\"\n",
    "    show_log = kwargs.get('show_log', True)\n",
    "    show_error = kwargs.get('show_error', False)\n",
    "    compresslevel = kwargs.get('compresslevel', 1)\n",
    "    threads = kwargs.get('threads', 4)\n",
    "    flag = False\n",
    "    # input: src\n",
    "    if not isinstance(src, str):\n",
    "        if show_error:\n",
    "            log.error('src expect str, got {}'.format(src))\n",
    "    if not file_exists(src):\n",
    "        if show_error:\n",
    "            log.error('src not exists, {}'.format(src))\n",
    "    # output: dest\n",
    "    if dest is None:\n",
    "        dest = os.path.splitext(src)[0] if decompress else src + '.gz'\n",
    "    if isinstance(dest, str):\n",
    "        if file_exists(dest):\n",
    "            if show_error:\n",
    "                log.error('dest exists, {}'.format(dest))\n",
    "        elif os.path.exists(os.path.dirname(dest)):\n",
    "            flag = True\n",
    "        else:\n",
    "            if show_error:\n",
    "                log.error('dest not valid, {}'.format(dest))\n",
    "    else:\n",
    "        if show_error:\n",
    "            log.error('dest expect str, got {}'.format(dest))\n",
    "    # do-the-thing\n",
    "    out = None\n",
    "    if flag:\n",
    "        is_gzipped = file_is_gzipped(src)\n",
    "        if decompress:\n",
    "            if is_gzipped:\n",
    "                with xopen(src, 'rb') as r, \\\n",
    "                    xopen(dest, 'wb') as w:\n",
    "                    shutil.copyfileobj(r, w)\n",
    "                out = dest\n",
    "            else:\n",
    "                if show_error:\n",
    "                    log.error('src is not gzipped, skipped. {}'.format(src))\n",
    "        else:\n",
    "            if is_gzipped:\n",
    "                if show_error:\n",
    "                    log.error('src is gzipped, skipped. {}'.format(src))\n",
    "            else:\n",
    "                with xopen(src, 'rb') as r, \\\n",
    "                    xopen(dest, 'wb', threads=threads,\n",
    "                          compresslevel=compresslevel) as w:\n",
    "                    shutil.copyfileobj(r, w)\n",
    "                out = dest\n",
    "    return out\n",
    "\n",
    "\n",
    "\n",
    "x = 'test/data/pe_control_rep1_1.fq.gz'\n",
    "y = 'test/aaa.fq'\n",
    "\n",
    "gzip_file(x, y)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "driven-grade",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-03-18T11:36:27.674223Z",
     "iopub.status.busy": "2021-03-18T11:36:27.673692Z",
     "iopub.status.idle": "2021-03-18T11:36:27.817500Z",
     "shell.execute_reply": "2021-03-18T11:36:27.816968Z",
     "shell.execute_reply.started": "2021-03-18T11:36:27.674171Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function xopen in module xopen:\n",
      "\n",
      "xopen(filename, mode: str = 'r', compresslevel: Union[int, NoneType] = None, threads: Union[int, NoneType] = None) -> <class 'IO'>\n",
      "    A replacement for the \"open\" function that can also read and write\n",
      "    compressed files transparently. The supported compression formats are gzip,\n",
      "    bzip2 and xz. If the filename is '-', standard output (mode 'w') or\n",
      "    standard input (mode 'r') is returned.\n",
      "    \n",
      "    The file type is determined based on the filename: .gz is gzip, .bz2 is bzip2, .xz is\n",
      "    xz/lzma and no compression assumed otherwise.\n",
      "    \n",
      "    mode can be: 'rt', 'rb', 'at', 'ab', 'wt', or 'wb'. Also, the 't' can be omitted,\n",
      "    so instead of 'rt', 'wt' and 'at', the abbreviations 'r', 'w' and 'a' can be used.\n",
      "    \n",
      "    In Python 2, the 't' and 'b' characters are ignored.\n",
      "    \n",
      "    Append mode ('a', 'at', 'ab') is not available with BZ2 compression and\n",
      "    will raise an error.\n",
      "    \n",
      "    compresslevel is the compression level for writing to gzip files.\n",
      "    This parameter is ignored for the other compression formats. If set to\n",
      "    None (default), level 6 is used.\n",
      "    \n",
      "    threads only has a meaning when reading or writing gzip files.\n",
      "    \n",
      "    When threads is None (the default), reading or writing a gzip file is done with a pigz\n",
      "    (parallel gzip) subprocess if possible. See PipedGzipWriter and PipedGzipReader.\n",
      "    \n",
      "    When threads = 0, no subprocess is used.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from xopen import xopen\n",
    "\n",
    "help(xopen)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
